MVP

juanimedone opened the pull request:
# MVP
## Summary
This pull request evolves the project from scratch to a basic implementation of Conway's Game of Life. It includes the game logic, visual rendering, and some utilities for validation and testing.

## Changes
### Game Logic Implementation
- Added the `Game` struct to represent the state of the game.
- Implemented the core game functions:
  - `new`: Initializes a new game with a **random** state.
  - `get_index`: Computes the index for a given cell position ((x,y) coordinates).
  - `count_neighbors`: Counts the number of live neighbors for a given cell.
  - `update`: Applies the rules of the game to update the grid.
  - `draw`: Renders the game state to the screen.

### Game Initialization and Rendering
- Created a `start` function to handle game initialization and rendering using the `macroquad` library.
- Added parameter validation in `validate_start_parameters` to ensure valid game settings.

### Testing
- Added unit tests for core game functions, including:
   - Validating start parameters.
   - Verifying index calculations.
   - Testing neighbor counting and game state updates.
- These include handling edge cases to ensure correct functionality.
- Fixed Clippy warnings and applied `cargo fmt` for code consistency and to prevent future bugs.

juanimedone committed:
Add basic game functionality

1- Creates a random initial state of the cells.
2- Updates them taking into account their neighbors.
3- Draws the updated grid.
4- Goes back to step 2 in an infinite loop.

juanimedone committed:
Add Execution and Rules section to README.md

juanimedone committed:
Move constants outside of game.rs

Both height and width are now being kept as game's state.
cell_size is only used to draw.

juanimedone committed:
Move next_cells from Game attributes to fn update's scope

juanimedone committed:
Add height, width and cell_size validations

juanimedone committed:
Add tests and refactor functions

- Simplified get_index by removing unnecesary bound wrapping.
- count_neighbors checks that neither x or y are out of bounds.
- Extracted new function validate_start_parameters from start
- Added tests for validate_start_parameters, (Game) new, get_index, count_neighbors and update

juanimedone committed:
Add documentation

juanimedone committed:
Add documentation to main.rs and lib.rs

juanimedone committed:
Fix Clippy warning and run cargo fmt

juanimedone committed:
Add execution comment  in README.md

MarcosNicolau commented on src/game.rs:
```diff
@@ -0,0 +1,250 @@
+struct Game {
+    height: usize,
+    width: usize,
+    cells: Vec<bool>,
```
Why don't you use a matrix instead?

vicentevieytes replied:
Agree, it'd be more readable with a Vec<Vec\<Bool> and indexing would be more intuitive.

juanimedone replied:
You're right, I'll change that.

MarcosNicolau commented on src/game.rs:
```diff
@@ -0,0 +1,250 @@
+    /// # Returns
+    ///
+    /// The index in the cells vector.
+    fn get_index(&self, x: i32, y: i32) -> usize {
```
You would avoid doing this calculations for example

MarcosNicolau commented on src/game.rs:
```diff
@@ -0,0 +1,250 @@
+    fn draw(&self, cell_size: f32) {
+        for x in 0..self.width {
+            for y in 0..self.height {
+                let index = y * self.width + x;
```
shouldn't you use the `get_index` here?

juanimedone replied:
Yes, I changed `get_index`'s original code and forgot to update this part. Anyways, I'll change `cells` to be a `Vec<Vec<bool>>` so that indexing function won't be needed anymore.

MarcosNicolau commented on src/game.rs:
```diff
@@ -0,0 +1,250 @@
+/// # Returns
+///
+/// A message if an error ocurrs while validating the parameters, otherwise it loops infinitely.
+pub async fn start(height: usize, width: usize, cell_size: f32) -> Result<(), String> {
```
I would put the game loop inside the `Game` struct, since they are closely related and, indeed, it is the whole point of the game.

juanimedone replied:
Ok, will do.

MarcosNicolau commented on src/game.rs:
```diff
@@ -0,0 +1,250 @@
+    ///
+    /// A new `Game` instance.
+    fn new(height: usize, width: usize) -> Self {
+        let cells = (0..height * width).map(|_| rand::random()).collect();
```
aren't you creating a cell for every pixel here? shouldn't you divide the grid size by the cell size?

vicentevieytes replied:
This seems correct to me.

new just initializes the grid with the amount of elements indicated by height and width, then drawing is performed by the draw function that takes the cell size and multiplies it for every element in the grid.

MarcosNicolau replied:
Yeah sure, but what I meant is: aren't you creating more cells than needed?

MarcosNicolau replied:
For instance, if the screen is 1024x800 -> size(cells) = 819.200. Now if your cells size is, say, 5, then size(cells) = 1024/5 * 800/5 = 32.768

vicentevieytes replied:
I'm not sure honestly. Maybe if there are more cells than they fit in the screen there could be some issues with the way the program draws to the screen now that i think about it.

vicentevieytes replied:
> For instance, if the screen is 1024x800 -> size(cells) = 819.200. Now if your cells size is, say, 5, then size(cells) = 1024/5 * 800/5 = 32.768

The height and width passed to the `new` function are not necessarily the screen height and width. They are hard coded in main.rs in this case.

MarcosNicolau replied:
Imagine you pass the screen size as params, but it applies to any value

vicentevieytes replied:
Yeah, then it does get messed up i just tried it. But if you pass sreen-size/cell-size as params it works alright

juanimedone replied:
Didn't notice that, I'll fix it.

MarcosNicolau reviewed (changes requested):
The game works fine, try to address the comments, merge it, and then add some new features, for example:

- Draw grid lines
- Add a start menu
- Ability to pause the game
- Ability to choose the initial alive seeds through a GUI
- Ability to change the game speed

vicentevieytes reviewed (changes requested):
I think you should change the cells matrix from being a single array to an explicit matrix like Vec<Vec\<bool>. That you have to use get_index every time you want to index in the matrix seems unnecessarily complex to me.

The rest of the code seems correct to me, but i think you should make this change now before a lot more code is added to the project.

juanimedone committed:
Change cells representation and remove get_index

- cells turns to a matrix for clearer indexing.
- get_index function no longer needed.
- Updated related tests.

juanimedone committed:
Move start and validate_start_parameters inside Game

- Modify main's entry point.
- Keep cell_size as game's state.
- Update tests and doc.

juanimedone committed:
Fix grid size

- Change number of rows and cols to be divided by cell_size.
- Remove height and width from game's state.

juanimedone commented:
I made several changes based on the comments you provided. These include:
- Turning `cells` to a matrix (`Vec<Vec<bool>>`) [like you proposed](https://github.com/juanimedone/conways-game-of-life/pull/1#discussion_r1684944107) in order to get more understandability and to avoid unnecessary calculations.
- Moving the game's main loop (`start` function) inside `Game` as a struct method. Refers to this [comment](https://github.com/juanimedone/conways-game-of-life/pull/1#discussion_r1684946951).
- Fixing the [grid size issue](https://github.com/juanimedone/conways-game-of-life/pull/1#discussion_r1684950525).

Please check them out and if everything looks fine for you I would suggest merging this PR and start working on the next features.

vicentevieytes reviewed (approved):
🚀💫Looks good

MarcosNicolau reviewed (approved):
LGTM 👍
